### ISSUE TITLE
intents originating from components resp. scripts

### ISSUE THREAD

**Issue by h0ru5** [w3c/editing#16]
a use case that could be interesting would be that components or scripts can issue intents, thus keeping the actual behavior to fulfill the intent compartmentized to the final application.

E.g.in a map view on a dashboard displaying traffic should zoom and scroll to the location if an accident happens and can optionally notify etc.

**Comment by BenjamP** [comment]
I'm not sure I follow this. Do you mind giving more details or examples?

**Comment by h0ru5** [comment]
Hmm.. I tried writing it up in the PR. Its not too much about editing as such, but more about the intent system you propose. Let me try to rephrase it:
So far, intents are described as actions that a user did resp. meant, directed from the user to a component. However, I see use cases where the originator is not the user or an assistive technology in his stead, but a component, as well as the target could be the user or another component.
Examples would be (push) notifications originating from the backend, or interactions between components such as a listbox of locations and a map.

I think a common method of abstraction would help decoupling the components and the proposed intents seem to be a very fitting candidate IMHO.

**Comment by BenjamP** [comment]
Interesting idea. I think we should use intents (Device Independent Events) for direct user input only. A backend event should just direct call the functionality it needs. Say for instance there was a bold() function in a site's script. The site should listen for bold beforeInput events and call their bold() inside the event. If the backend decided to make text bold (say from another concurrent user), the site should just directly call bold(), not fire a bold beforeInput event. The goal is to simplify various types of user input in a single event, not to replace function().

Does that make sense?

**Comment by h0ru5** [comment]
tl;dr: yes. you're right. I might be chasing the dragon called "make everything generic".

My goal was to abstract and decouple, like in the example the front-end and the backends 
e.g. save()-function called from the editor does not need to know how the backend will save it. (e.g. onto ondrive, dropbox, owncloud or just offer it as a download).
This will require some kind of contract for specific cases and optionally an generic abstraction.

In the web dev world,  I would expect registering a callback with the editor to achieve this. The idea was to define/using intents quite similar to your proposal for that. Advantage might be they can be passed  as messages between web workers or even web sockets. 

But I agree that is probably way further down the road.

**Comment by johanneswilm** [comment]
Indeed, lets keep intentions to things that are coming from the user. We can already call functions in Javascript by clicking on things. :)

The only reason we use intentions is because people have different input methods and we don't care about how they indicate a specific input intent.  For everything else there are functions.

CLosing for now. Please reopen if something has been missed.
